lib/http1.c

```diff
@@ -181,8 +175,13 @@ static void set_req_io_timeout(struct st_h2o_http1_conn_t *conn, uint64_t timeou
     if (conn->_io_timeout_entry.cb != NULL)
         h2o_timer_unlink(&conn->_io_timeout_entry);
     conn->_io_timeout_entry.cb = cb;
-    if (cb != NULL)
+    h2o_linklist_unlink(&conn->super._conns);
+    if (cb != NULL) {
         h2o_timer_link(conn->super.ctx->loop, timeout, &conn->_io_timeout_entry);
+        h2o_linklist_insert(&conn->super.ctx->_conns.idle, &conn->super._conns);
+    } else {
+        h2o_linklist_insert(&conn->super.ctx->_conns.active, &conn->super._conns);
+    }
```

Comment:
> cleanup_connection with http1_is_persistent set to true but is yet to receive a single byte

I'm having troubles re-producing this situation via test case, could you expand on this?

Reply:
Look at `handle_one_body_fragment`. This function calls `clear_timeouts` that invoke `set_req_io_timeout` with the last parameter set to NULL.

The logic here is to disable I/O timeout while the server processes a chunk of request body that has been received (which can take long time when the server is forwarding that chunk to the origin), then reactivate I/O timeout when the `proceed_request` callback is being called.

I believe we have similar logic on the send-side (i.e., on the path we send response).

To summarize, I/O timeout is enabled only when we the client is expected to make the next state change. It is turned off while the request is inflight but the server is the one to trigger the state change.

Reply:
My assumption is that `cleanup_connection` get's called when h2o has finished reading/processing the request & body and can start reading / sending the next request.

> but is yet to receive a single byte

In this case do you mean the response from the origin? Do we currently collect this information? I see `conn->req->res` but maybe I'm mis-understanding.

Reply:
> My assumption is that `cleanup_connection` get's called when h2o has finished reading/processing the request & body and can start reading / sending the next request.

That is correct.

>> but is yet to receive a single byte
> In this case do you mean the response from the origin?

No, I mean (yet to receive a single byte of) a request from client. But to be precise, let's discuss each state of a HTTP/1 connection.

Let's look at each state.
* When the client connects to the server using HTTP/1, `io_timeout` is used, and the connection is *not* in idle state. The belief here is that client is connecting because it is going to send a request.
* When there is something inflight to the client, `io_timeout` is used, because we expect the client to make progress within that timeout as well.
* When the ball is at the server side but the server does not have anything to send (waiting for upstream to supply more data), timeout is disabled (i.e., the callback is set to NULL).
* Once all response have been sent and the connection is to persist, we adopt `idle_timeout`, because at this moment, we do not know if the client is going to immediately send a new request on that connection, or if the client is just going to hold onto the connection as it might want to use it in the future. Once we receive a byte, we know that the client is sending a request, hence we adopt `io_timeout`.

The observation behind this design is that "idle" is a moment when the server thinks a *client* might not have a request inflight. That's when we can cull the connection.

Reply:
Thanks for explaining!

I added a check to make sure the input socket is empty before going idle; do you believe we also need to check the `conn->_io_timeout_entry.cb`? In my testing this is non-null when `cleanup_connection` is called  
https://github.com/h2o/h2o/pull/2932/commits/4cd53680dc2faaed55c93442c0a93fa8673b21bd